src/server/pfs/fuse/server.go

```diff
@@ -0,0 +1,840 @@
+	// with their mount states
+	lr := map[string]RepoResponse{}
+	repos, err := mm.Client.ListRepo()
+	if err != nil {
+		return lr, err
+	}
+	for _, repo := range repos {
+		rr := RepoResponse{Name: repo.Repo.Name, Branches: map[string]BranchResponse{}}
+		bs, err := mm.Client.ListBranch(repo.Repo.Name)
+		if err != nil {
+			return lr, err
+		}
+		for _, branch := range bs {
+			br := BranchResponse{Name: branch.Branch.Name}
+			k := MountKey{
+				Repo:   repo.Repo.Name,
+				Branch: branch.Branch.Name,
+				Commit: "",
+			}
+			s, ok := mm.States[k]
+			if ok {
+				br.Mount = s.MountState
+			} else {
+				br.Mount = MountState{State: "unmounted"}
+			}
+			rr.Branches[branch.Branch.Name] = br
+		}
+		lr[repo.Repo.Name] = rr
+	}
+
+	// TODO: also add any repos/branches that have been deleted from pachyderm
+	// but are still mounted here :-O we should send them a special signal to
+	// tell them they're "stranded" or "missing" probably?
+
+	return lr, nil
+}
+
+func NewMountStateMachine(key MountKey, mm *MountManager) *MountStateMachine {
+	return &MountStateMachine{
+		MountState: MountState{
+			MountKey: key,
+		},
+		manager:   mm,
+		mu:        sync.Mutex{},
+		requests:  make(chan Request),
+		responses: make(chan Response),
+	}
+}
+
+func (mm *MountManager) MaybeStartFsm(key MountKey) {
+	mm.mu.Lock()
+	defer mm.mu.Unlock()
+	_, ok := mm.States[key]
+	if !ok {
+		// set minimal state and kick it off!
+		m := NewMountStateMachine(key, mm)
+		mm.States[key] = m
+		go func() {
+			m.Run()
+			// when execution completes, remove ourselves from the map so that
+			// we can infer from the key in the map that the state machine is
+			// running
+			mm.mu.Lock()
+			defer mm.mu.Unlock()
+			delete(mm.States, key)
+		}()
+	} // else: fsm was already running
+}
+
+func (mm *MountManager) MountBranch(key MountKey, name, mode string) (Response, error) {
+	// name: an optional name for the mount, corresponds to where on the
+	// filesystem the repo actually gets mounted. e.g. /pfs/{name}
+	mm.MaybeStartFsm(key)
+	mm.States[key].requests <- Request{
+		Mount:  true,
+		Repo:   key.Repo,
+		Branch: key.Branch,
+		Commit: key.Commit,
+		Name:   name,
+		Mode:   mode,
+	}
+	response := <-mm.States[key].responses
+	return response, response.Error
+}
+
+func (mm *MountManager) UnmountBranch(key MountKey, name string) (Response, error) {
+	mm.MaybeStartFsm(key)
+	mm.States[key].requests <- Request{
+		Mount:  false,
+		Repo:   key.Repo,
+		Branch: key.Branch,
+		Commit: key.Commit,
+		Name:   name,
+		Mode:   "",
+	}
+	response := <-mm.States[key].responses
+	return response, response.Error
+}
+
+func NewMountManager(c *client.APIClient, target string, opts *Options) (ret *MountManager, retErr error) {
+	if err := opts.validate(c); err != nil {
+		return nil, err
+	}
+	commits := make(map[string]string)
+	for repo, branch := range opts.getBranches() {
+		if uuid.IsUUIDWithoutDashes(branch) {
+			commits[repo] = branch
```

Comment:
What is `commits` being used for? There doesn't seem to be a reference to it afterwards.

Reply:
Yeah I can't quickly tell either?